grammar/jbuild

```diff
@@ -0,0 +1,28 @@
+(jbuild_version 1)
+
+(library
+ ((name grammar)
+  (public_name coq.grammar)
+  (wrapped false)
+  (flags (-rectypes))
+  (libraries (camlp5))))
+
+(rule
+ ((targets (q_util.ml))
+  (deps (q_util.mlp))
+  (action (system "camlp5o pa_extend.cmo q_MLast.cmo pa_macro.cmo pr_o.cmo -loc loc -impl ${<} -o ${@}"))))
```

Comment:
Why system? do you need the shell for something here?

`(run ${bin:camlp5o} pa_extend.cmo ... ${<} -o ${@})` should suffice.

Reply:
Thanks !

Reply:
By the way when this matures a bit more I was planning to submit a "status report" to dune devs, do you think that would be useful?

Reply:
Yeah, that would be very useful. We'd like to get as much feedback as possible before 1.0

Reply:
Cool. But at first sight, I think we don't have any new or surprising feedback over all the existing discussion and feature requests.

Reply:
> (run ${bin:camlp5o} pa_extend.cmo ... ${<} -o ${@}) should suffice.

Note that I cannot use this as it will pick `camlp5o.opt` which doesn't work as we really need the bytecode version as to load the corresponding plugins.

I guess this was done to support architectures without a native backend, but I don't know how to make `camlp5o` work in native mode while loading all the corresponding extensions.

Reply:
NP. I think there's definitely some pain points that you could bring up however. For example, all this duplication with camlp5. I think globs alone wouldn't be enough to address it either, since it's spread across dirs.

Perhaps another interesting avenue to pursue would be to build coq projects themselves in dune. Writing rules in OCaml is a lot simpler than Makefiles. Though that might have to wait for dune's upcoming internal API change.

Reply:
> Note that I cannot use this as it will pick camlp5o.opt which doesn't work as we really need the bytecode version as to load the corresponding plugins.

I see. And what about `(run camlp5o ...)`. I think that shouldn't try to be smart.

Reply:
`(run camlp5o ...)` also selects the `opt` version. Yeah, Coq has some quite special preprocessing needs. That being said if we could build a custom static `camlp5o` it would speed up the compilation by a few seconds.



> all this duplication with camlp5. I think globs alone wouldn't be enough to address it either, since it's spread across dirs.

What we want here is a custom rule to pre-process `ml4` files. I find the `ml4` extension convenient as it tags the files that need the preprocessor. Coq may abandon `camlp5` in favor of `ppx` soon, however we'd still want a convenient form to select files that need preprocessing. Applying the preprocessor to all files is not an option, writing it manually is painful and prone to errors.

> Perhaps another interesting avenue to pursue would be to build coq projects themselves in dune.

This is in the roadmap.

Reply:
> This is in the roadmap.

I think you mean your personal roadmap. Because AFAIK nothing like this build system change was announced so far.

Reply:
In the PR roadmap.

Reply:
Efforts on using Ocamlbuild and now Dune have been going on for years, first started by Pierre, this PR in particular has been mentioned many many times in many different contexts....

Reply:
Last time it was mentioned was in the WG discussion of Monday about interfaces, but yeah, maybe that was one of the connections that broke.

Reply:
By the way I avoided the problem of `Bin.which` always selecting the `opt` version by actually building a custom `coqp5`, which works much better. In terms of `ml4 -> ml` preprocessing a kind of `ocamllex` rule would suffice for us, as dependencies are 1:1. For `.v -> .vo` I just used `coqdep` to generate the `jbuild` files containing the proper `(rule ..)` and `(install ...)` stanzas.

The whole system builds and installs beautifully.